### PR TITLE
Trigger symfloat specialization in argument binding code

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139933
* #139583
* #138922
* #138918
* #138666
* #138915
* #139587
* #139935
* #139896
* __->__ #139454
* #139846
* #139572
* #139568
* #139457
* #139569

Fixes the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:2390]: Merge this with below? If you can unconditionally let this through that would be good too lol

File: torch/nn/functional.py
Comment (ezyang) [inline torch/nn/functional.py:4642]: This is a suspicious spot to do it, would have expected later

File: torch/_dynamo/utils.py
Comment (ezyang) [inline torch/_dynamo/utils.py:2195]: Also a suspicious spot to do it

File: torch/nn/functional.py
Comment (bobrenjc93) [reply torch/nn/functional.py:4642]: Did you have a place in mind? Generally for all of these fixes I'm specializing right before the point it needs to in the stacktrace (via debugger + jumping up and down stacks). If there's a better method to find the precise place to put these specializations, do let me know!

File: torch/_dynamo/utils.py
Comment (bobrenjc93) [reply torch/_dynamo/utils.py:2195]: Did you have a place in mind? Generally for all of these fixes I'm specializing right before the point it needs to in the stacktrace (via debugger + jumping up and down stacks). If there's a better method to find the precise place to put these specializations, do let me know!

Comment (ezyang) [reply torch/_dynamo/utils.py:2195]: What I am generically expecting is if I pass a SymFloat to a PyTorch API that only accepts 'float', I expect that API to automatically trigger specialization in its argument binding code (as it attempts to get a float and gets a SymFloat instead.) The analogous code for SymInt was added in https://github.com/pytorch/pytorch/pull/95479 so you probably need the float version of it too.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp:37]: All of this code for SymFloat and SymFloatArrayRef isn't actually doing anything, because you are never code generating SymFloat/SymFloatArrayRef entries in the string passed to Python argument parser.  You can drop all of it, you only need to change the semantics of `is_float_or_complex_list` to also accept SymFloat, and ditto with THPUtils_checkDouble to switch to is_float_or_symfloat.

**Review by ezyang — APPROVED** [review]
I'm approving because the bogus changes are harmless, but it is better to excise them before landing.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (bobrenjc93) [reply torch/csrc/utils/python_arg_parser.cpp]: Should we do the same thing for SymInt? If not, what's the difference?

**Pull request closed without merging**
